csrc/device_lower/analysis/thread_predicate.cpp

```diff
@@ -494,6 +503,221 @@ class RedundantUseAnalysis : BackwardVisitor {
+class ConcretizedBroadcastRedundantWriteRemover {
+ public:
+  // interface to run the check
+  ConcretizedBroadcastRedundantWriteRemover(const TensorView* out_tv)
+      : tv_(out_tv), root_domain_(out_tv->getMaybeRFactorDomain()) {
+    setCandidateLeafDomains();
+    if (!candidate_leaf_domains_.empty()) {
+      setConcretizedBroadcastRootDomain();
+      if (!concretized_broadcast_root_domains_.empty()) {
+        for (auto ld : candidate_leaf_domains_) {
+          // find all root domains that are merged to this leaf domain.
+          const std::vector<IterDomain*>& merged_root_domains =
+              getRootDomainsMergedToLeaf(ld);
+          if (!merged_root_domains.empty()) {
+            const ParallelType& pt = ld->getParallelType();
+            Val* write_index =
+                getIndexWithoutBroadcast(merged_root_domains, pt);
+            write_index_map_[pt] = write_index;
+          }
+        }
+      }
+    }
+  }
+  // interface to get results
+  std::unordered_map<ParallelType, Val*> getWriteIndexMap() {
+    return write_index_map_;
+  }
+
+ private:
+  const TensorView* tv_;
+  const std::vector<IterDomain*>& root_domain_;
+  // leaf domains that are merged from root domains and parallelized by thread
+  // blocks
+  std::vector<IterDomain*> candidate_leaf_domains_;
+  // map from root domain to its concretized domain
+  std::unordered_map<IterDomain*, IterDomain*>
+      concretized_broadcast_root_domains_;
+  // map from parallel type to its write index
+  std::unordered_map<ParallelType, Val*> write_index_map_;
+
+  void setCandidateLeafDomains() {
+    for (auto ld : tv_->domain()->leaf()) {
+      const ParallelType& pt = ld->getParallelType();
+      auto merge = dynamic_cast<Merge*>(ld->definition());
+      if (isParallelTypeThread(pt) && merge) {
+        candidate_leaf_domains_.push_back(ld);
+      }
+    }
+  }
+
+  // Current implementation only works for broadcast root domain uniquely
+  // concretized. Will skip if concretized to differet domains, and usually
+  // should be segmented, see
+  // FusionAvoidRedundantWriteDifferentConcretizedDomains_CUDA
+  void setConcretizedBroadcastRootDomain() {
+    for (auto rd : root_domain_) {
+      if (rd->isBroadcast()) {
+        IterDomain* identical_concretized_domain = nullptr;
+        // find this broadcast root domain's identical concretized domain
+        // from its exact map (includes itself).
+        const auto& mapped_rd_exact_set = GpuLower::current()
+                                              ->caMap()
+                                              ->getIdSets(IdMappingMode::EXACT)
+                                              .getDisjointSetOf(rd);
+        for (auto mapped_rd : mapped_rd_exact_set) {
+          std::unordered_set<IterDomain*> all_cids =
+              GpuLower::current()
+                  ->concretizedBroadcastDomains()
+                  ->allConcretizedDomains(mapped_rd);
+
+          if (!all_cids.empty()) {
```

Comment:
There are many broadcast domains that are not concretized. We only need to make sure all the concretized domains are same.
```
        // skip if this mapped domain is not concretized
        // don't need to reset identical_concretized_domain
        // since we only needs to make sure all the concretized domains are same
        if (all_cids.empty()) {
          continue;
        }
```

Reply:
Not sure why. Don't we only support concretized broadcast domains? Why can we ignore non-concretized domains?

Reply:
Because this loop is for the broadcast root domain itelsef and its exact mapped domains. e.g. in case `FusionAvoidRedundantWriteNonOutput_CUDA`, the kernel is:
```
T2_l[ iS3{i0} ] ca_pos( 1 )
   = Set( T0_g[ iS0{i0} ] )
T3_l[ iblockIdx.x12{( i0 * 1 )} ] ca_pos( 1 ) produce_pos( 1 )
   = broadcast( T2_l[ iS3{i0} ] ca_pos( 1 ) )
T4_g[ iblockIdx.x13{( i0 * i3 )} ] produce_pos( 1 )
   = **T3_l**[ iblockIdx.x12{( i0 * 1 )} ] ca_pos( 1 ) produce_pos( 1 )
   + T1_g[ iS1{i2}, iS2{i3} ];
**T5_g**[ iblockIdx.x14{( i0 * 1 )} ] ca_pos( 1 ) produce_pos( 1 )
   = T3_l[ iblockIdx.x12{( i0 * 1 )} ] ca_pos( 1 ) produce_pos( 1 )
   + double(1);
**T6_g**[ iblockIdx.x15{( i0 * 1 )} ] produce_pos( 1 )
   = T5_g[ iblockIdx.x14{( i0 * 1 )} ] ca_pos( 1 ) produce_pos( 1 )
   + double(1);
```
The broadcast root domain of `T3_l` is concretized but `T5_g` and `T6_g` are not.
![image](https://github.com/NVIDIA/Fuser/assets/116412316/b842e3b5-0c20-4cfb-b879-e434cb94de6f)

Almost all the added test cases in this PR include non-concretized broadcast root domains.

Reply:
Aren't we assuming all the exactly mapped IDs are uniquely concretized to the same ID? Isn't the case like the above violating the assumption? 

Note that `mapped_rd` can be `rd` itself. Is it valid to just skip if `rd` isn't concretized at all but another exactly mapped broadcast ID is uniquely concretized?

Reply:
This is exactly what happened in the original issue, where `rd` itself is not concretized but one of its exactly mapped ID is concretized.
This is also why I need to search for concretized domain from all the exactly mapped domains instead of just the broadcast domain itself.

Reply:
Doesn't that mean the assumption is invalid then?

I think we need to look at the loop concrete ID because that dictates how the broadcast ID is indexed.

Reply:
when I say concretized in the assumption, I mean "itself or one of its exactly mapped ID is concretized." If we only consider the broadcast ID itself, then we can't solve the original issue.
What's "loop concrete ID"?

Reply:
(1) I tried `GpuLower::current()->caMap()->getConcreteMappedID(rd, IdMappingMode::LOOP);`
it doesn't work, because in some cases, e.g. `FusionAvoidRedundantWriteNonOutput_CUDA` we don't have loop.
(2) I also tried `GpuLower::current()->caMap()->getConcreteMappedID(rd, IdMappingMode::EXACT);`. Two cases failed in validation, `FusionAvoidRedundantWrite_CUDA` and `FusionAvoidRedundantWriteDifferentConcretizedDomains_CUDA`
I checked `runTest({true, false, true, false});
` in `FusionAvoidRedundantWrite_CUDA`, the broadcast domain -> concrete_id map is:
```
bS32{1} --> concrete_id_0 bS32{1}
bS34{1} --> concrete_id_0 bS34{1}
```
Not sure why `getConcreteMappedID` returns a `bS32{1}` which doesn't seem to be a concretized domain.
while the right one should be:
```
bS32{1} --> concrete_id_0 iS155{T1.size[0]}
bS34{1} --> concrete_id_0 iS157{T1.size[2]}
```
Code I used:
```
  void setConcretizedBroadcastRootDomain() {
    for (auto rd : root_domain_) {
      if (!rd->isBroadcast()) {
        continue;
      }
      auto concrete_id_0 =
        GpuLower::current()->caMap()->getConcreteMappedID(rd, IdMappingMode::EXACT);
      if (concrete_id_0) {
        concretized_broadcast_root_domains_[rd] = concrete_id_0;
      }
    }
  }
```

Reply:
Looks like the current implementation is the most robust. what' the potential issue?

Reply:
> Not sure why getConcreteMappedID returns a bS32{1} which doesn't seem to be a concretized domain.

Please study what it does. It's going to be replaced with #32, but the current code may be easier to start with.

Reply:
> (1) I tried GpuLower::current()->caMap()->getConcreteMappedID(rd, IdMappingMode::LOOP);
it doesn't work, because in some cases, e.g. FusionAvoidRedundantWriteNonOutput_CUDA we don't have loop.

How did you try? What does the error look like? I don't understand what there's no loop.

Reply:
1. How did you try?
I replaced original `setConcretizedBroadcastRootDomain` with 
```
  void setConcretizedBroadcastRootDomain() {
    for (auto rd : root_domain_) {
      if (!rd->isBroadcast()) {
        continue;
      }
      auto concrete_id_0 =
        GpuLower::current()->caMap()->getConcreteMappedID(rd, IdMappingMode::LOOP);
      if (concrete_id_0) {
        std::cout << "rd " << rd->toString() << "concrete_id_0 " << concrete_id_0->toString() << std::endl;
        concretized_broadcast_root_domains_[rd] = concrete_id_0;
      }else{
        std::cout << "rd " << rd->toString() << "concrete_id_0 is null" << std::endl;
      }
    }

  }
```
2. What does the error look like?
The redundant write is not removed.
`C++ exception with description "TV5 and TV6 should be predicted by ParallelType::BIDx with a write_index_map!`

3. " I don't understand what there's no loop."
I mean `      auto concrete_id_0 =
        GpuLower::current()->caMap()->getConcreteMappedID(rd, IdMappingMode::LOOP);` returns nullptr.
```
disjoint_set_shared_ptr not found in map for bS9{1}
rd bS9{1}concrete_id_0 is null
disjoint_set_shared_ptr not found in map for bS11{1}
rd bS11{1}concrete_id_0 is null
```
Here is the fusion_ir:
```
%kernel {
T2_l[ iS3{i0} ] ca_pos( 1 )
   = Set( T0_g[ iS0{i0} ] )
T3_l[ iblockIdx.x12{( i0 * 1 )} ] ca_pos( 1 ) produce_pos( 1 )
   = broadcast( T2_l[ iS3{i0} ] ca_pos( 1 ) )
T4_g[ iblockIdx.x13{( i0 * i3 )} ] produce_pos( 1 )
   = T3_l[ iblockIdx.x12{( i0 * 1 )} ] ca_pos( 1 ) produce_pos( 1 )
   + T1_g[ iS1{i2}, iS2{i3} ];
T5_g[ iblockIdx.x14{( i0 * 1 )} ] ca_pos( 1 ) produce_pos( 1 )
   = T3_l[ iblockIdx.x12{( i0 * 1 )} ] ca_pos( 1 ) produce_pos( 1 )
   + double(1);
T6_g[ iblockIdx.x15{( i0 * 1 )} ] produce_pos( 1 )
   = T5_g[ iblockIdx.x14{( i0 * 1 )} ] ca_pos( 1 ) produce_pos( 1 )
   + double(1);

TransformPrinter : 
T0_g[ iS0{i0} ]
 root domain : (iS0{i0})
 contiguity: f
 leaf domain : (iS0{i0})
T2_l[ iS3{i0} ] ca_pos( 1 )
 root domain : (iS3{i0})
 contiguity: t
 leaf domain : (iS3{i0})
T3_l[ iblockIdx.x12{( i0 * 1 )} ] ca_pos( 1 ) produce_pos( 1 )
 root domain : (iS4{i0}, bS5{1})
 contiguity: t n
  Merge: iS4{i0} and bS5{1} -> iblockIdx.x12{( i0 * 1 )}
 leaf domain : (iblockIdx.x12{( i0 * 1 )})
T1_g[ iS1{i2}, iS2{i3} ]
 root domain : (iS1{i2}, iS2{i3})
 contiguity: f f
 leaf domain : (iS1{i2}, iS2{i3})
T4_g[ iblockIdx.x13{( i0 * i3 )} ] produce_pos( 1 )
 root domain : (iS6{i0}, iS7{i3})
 contiguity: t t
  Merge: iS6{i0} and iS7{i3} -> iblockIdx.x13{( i0 * i3 )}
 leaf domain : (iblockIdx.x13{( i0 * i3 )})
T5_g[ iblockIdx.x14{( i0 * 1 )} ] ca_pos( 1 ) produce_pos( 1 )
 root domain : (iS8{i0}, bS9{1})
 contiguity: t n
  Merge: iS8{i0} and bS9{1} -> iblockIdx.x14{( i0 * 1 )}
 leaf domain : (iblockIdx.x14{( i0 * 1 )})
T6_g[ iblockIdx.x15{( i0 * 1 )} ] produce_pos( 1 )
 root domain : (iS10{i0}, bS11{1})
 contiguity: t n
  Merge: iS10{i0} and bS11{1} -> iblockIdx.x15{( i0 * 1 )}
 leaf domain : (iblockIdx.x15{( i0 * 1 )})
}
```

Reply:
> > Not sure why getConcreteMappedID returns a bS32{1} which doesn't seem to be a concretized domain.
> 
> Please study what it does. It's going to be replaced with #32, but the current code may be easier to start with.

This is because when build the `concrete_id_cache_` for `exactNodes`, we just picked the first ID without checking whether it is a concretized ID or broadcast ID.
```
void ComputeAtMap::buildConcreteIds() {
  // For the exact map just select the first ID since they're all exactly the
  // same size, it doesn't matter which is selected. This should be run-to-run
  // deterministic but which ID gets selected her depends on the traversal order
  // generating the set (compute at map build).
  for (const auto& disjoint_set_shared_ptr :
       id_graph_.exactNodes().disjointSets()) {
    TORCH_INTERNAL_ASSERT(
        !disjoint_set_shared_ptr->vector().empty(),
        "Cannot compute concrete id of empty set.");
    auto first_id = disjoint_set_shared_ptr->vector().front();
    concrete_id_cache_[disjoint_set_shared_ptr] = first_id;
  }
```

Reply:
use `IdMappingMode::PERMISSIVE` works fine.

Reply:
> 1. How did you try?
>    I replaced original `setConcretizedBroadcastRootDomain` with
> 
> ```
>   void setConcretizedBroadcastRootDomain() {
>     for (auto rd : root_domain_) {
>       if (!rd->isBroadcast()) {
>         continue;
>       }
>       auto concrete_id_0 =
>         GpuLower::current()->caMap()->getConcreteMappedID(rd, IdMappingMode::LOOP);
>       if (concrete_id_0) {
>         std::cout << "rd " << rd->toString() << "concrete_id_0 " << concrete_id_0->toString() << std::endl;
>         concretized_broadcast_root_domains_[rd] = concrete_id_0;
>       }else{
>         std::cout << "rd " << rd->toString() << "concrete_id_0 is null" << std::endl;
>       }
>     }
> 
>   }
> ```


As mentioned in compute_at_map.h, the LOOP map is only valid for leaf IDs. I didn't expect it would return nullptr for non-leaf domains, which seems wrong. Do you have a small repro?

The LOOP map tells the actual IterDomain that defines the loop where a tensor is computed. Looks like you're trying to compute the actual concretized extent of a broadcast root domain, so you could do something like:

````
for (auto leaf_id : candidate_leaf_ids) {
  auto loop_concrete_id = caMap()->getConcreteMappedID(leaf_id, IdMappingMode::LOOP);
  auto concrete_root_ids = getInputsTo({loop_concrete_id});
  for (auto rd: root_domain_) {
    auto it = std::find_if(concrete_root_ids.begin(), concrete_root_ids.end(),
                      [] (auto concrete_ root_id) { return caMap()->areMapped(rd, concrete_root_id, 
  IdMappingMode::PERMISSIVE); });
    if (it == concrete_root_ids.end()) {
       // Failed to find the concrete ID. This could happen in complex broadcast and computeAt patterns. Not addressed for now
       continue;
     }
     auto concrete_root_id = *it;
     ...
  }
}
```